src/evo/creditpool.cpp

```diff
@@ -0,0 +1,334 @@
+// Copyright (c) 2023 The Dash Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <evo/creditpool.h>
+
+#include <evo/assetlocktx.h>
+#include <evo/cbtx.h>
+
+#include <llmq/utils.h>
+
+#include <chain.h>
+#include <logging.h>
+#include <validation.h>
+
+#include <algorithm>
+#include <exception>
+#include <memory>
+#include <stack>
+
+static const std::string DB_CREDITPOOL_SNAPSHOT = "cpm_S";
+
+std::unique_ptr<CCreditPoolManager> creditPoolManager;
+
+static bool getDataFromUnlockTx(const CTransaction& tx, CAmount& toUnlock, uint64_t& index, TxValidationState& state) {
+    CAssetUnlockPayload assetUnlockTx;
+    if (!GetTxPayload(tx, assetUnlockTx)) {
+        return state.Invalid(TxValidationResult::TX_CONSENSUS, "failed-creditpool-unlock-payload");
+    }
+
+    index = assetUnlockTx.getIndex();
+    toUnlock = assetUnlockTx.getFee();
```

Comment:
`getFee()` is uint32, so, it's >= 0 and always in range. DIP says nothing about minimum fee so far as I remember. @shumkov  ?

Reply:
ping re min fee

Reply:
as I talked with @shumkov: if an asset-unlock transaction will not be mined by any reason and timeout will happen - this transaction is changing state to "EXPIRY" and it is a terminal state for transaction.
In this case platform may generate new transaction for taking away credits from platform to blockchain (by new asset unlock transaction). New transaction can have higher fee to make miners be more interested in including this transaction in block.

Asset-unlock transaction has no inputs, but have outputs + spare field "fee" (getFee).
1. Good to check that all outputs are above dust value - because it's just regular output of tx.
2. this fee is not used separately, but sum-up will values of all other transaction and are used as part of "block reward". Even if transaction will have a fee 0.00000001 it won't be dust, because this fee will sum up with other values. 
3. there's validation of fee: `Consensus::CheckTxInputs` - this function setup value of `txfee` correctly and this fee checked again mempool limits, etc.

Beside that,  if any miner will include transaction with fee below `minRelayFee` - it doesn't mean that transaction is invalid. It is just miner decided to take tx with low fee, not validation's business - tx is valid.

Reply:
Thinking more about fees and stuff… we should change `feature_asset_locks.py` to avoid mining and creating/sending txes on the same node i.e. it should be 5 nodes total: 1. controller/mining node 2. wallet node 3,4,5. MNs. Currently we create and mine txes on the very same node (`self.nodes[0]`) so we don’t really see if smth stops these txes from being relayed or not. Lock txes should be created on the wallet node and unlock txes should be created on one of MNs. Also would be nice to see a test for unlock txes with different `fee`s (from 0 to >0.1 DASH).

Reply:
Why do we need to bypass limit? If fee is too low and tx is not mined - platform will generate new one after expiration. There's nothing special about asset-unlock.
 Tests works as I expected when implemented it...

For some txes such as quorum/etc there are nowhere to take reward/fee, so, checks are bypassed. But unlock txes are not special in this terms.
If fee is too small - it's okay to request "more fee, please"